compiler/rustc_mir_transform/src/instsimplify.rs

```diff
@@ -68,6 +69,35 @@ struct InstSimplifyContext<'a, 'tcx> {
 }
 
 impl<'tcx> InstSimplifyContext<'_, 'tcx> {
+    /// Transform aggregates like [0, 0, 0, 0, 0] into [0; 5].
+    /// GVN can also do this optimization, but GVN is only run at mir-opt-level 2 so having this in
+    /// InstSimplify helps unoptimized builds.
+    fn simplify_repeated_aggregate(&self, rvalue: &mut Rvalue<'tcx>) {
+        let Rvalue::Aggregate(box AggregateKind::Array(_), fields) = rvalue else {
+            return;
+        };
+        if fields.len() < 5 {
+            return;
+        }
+        let first = &fields[rustc_abi::FieldIdx::ZERO];
+        let Operand::Constant(first) = first else {
+            return;
+        };
+        let Ok(first_val) = first.const_.eval(self.tcx, self.typing_env, first.span) else {
+            return;
+        };
+        if fields.iter().all(|field| {
+            let Operand::Constant(field) = field else {
+                return false;
+            };
+            let field = field.const_.eval(self.tcx, self.typing_env, field.span);
```

Comment:
Would be cool to see more tests exercising this eval call. Specifically, maybe something like `[CONST, CONST, ...]` for some global const, something with ZSTs, something gauging whether or not it works with things like nested slices/string lits (probably doesn't? const eval doesn't deduplicate allocations like that right?)

Reply:
The optimization doesn't work on ZSTs. Ultimately that doesn't matter because RemoveZsts will run before InstSimplify and completely eliminate the statement we would optimize.

But it makes me a bit worried that this doesn't work with ZSTs because we generate a local per array element for ZSTs, but we don't for a numeric literal, and this optimization relies on not generating a local per array element.